README.md
# GeoBR

A geolocation library developed in Python to retrieve the user's location in Brazil.

## Download

https://github.com/uol/geo-br.git


## Prerequisites

The module depends on a KML file containing the coordinates of all Brazilian counties, which can be downloaded from:
http://www.gmapas.com/poligonos-ibge/municipios-do-brasil

Instaled virtualenv
```sh
pip install virtualenv
```

## Installing

Run within the project folder

Before start using the project you need converting the downloaded kml file into a binary file

```sh
python convert.py data.kml.gz

```

Create a virtual environment
```sh
make help           # show available options
make python-reqs    # install python packages in requirements.txt
make start          # launch a server from the local virtualenv
make test           # run the project tests
```

## Examples

```
curl -v "localhost:8080/?lat=-3.119856&lon=-60.045" -H "X-Token: fc0ae500c04c13425dc306c0342ebf85d6f260243c2bf0a9f6b7dbaa040db461";echo
```
```
HTTP/1.1 200 OK
Connection: close
Access-Control-Allow-Origin: *
X-Profile: 0
X-Req: 1ee3f430-43aa-49aa-b096-12861e4e90ff
X-Version: 2.1.0
X-Id: 127.0.0.1:44052
Content-Length: 43
Content-Type: application/json

{"city":"Manaus","state":"AM","found":true}
```

Response data

```
X-Profile : execution time in ms
X-Req : request ID
X-Version : module version
found: true => lat+long was found
found: false => lat+long not found, return the fallback result
```

Running the tests
Explain how to run the automated tests for this system

Give an example
And coding style tests
Explain what these tests test and why

Give an example
Deployment
Add additional notes about how to deploy this on a live system

## Built With
Sanic - The web framework used

## Contributing
Please read CONTRIBUTING.md for details on our code of conduct, and the process for submitting pull requests to us.

## Versioning
We use SemVer for versioning. For the versions available, see the tags on this repository.

## Authors
Ivan Rocha - Initial work

## License
This project is licensed under the GNU GENERAL PUBLIC LICENSE - see the LICENSE.md file for details

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.2] - 2019-06-13
### Changed
- Separação do teste de pesquisa para um arquivo de teste, usando o unittest

## [1.0.3] - 2019-07-04
### Changed
- Inclusão do Rio Grande do Sul no arquivo de conversão convert.py
- Inclusão de testes para certificar a existência de todos os estados e capitais brasileiras no kml

test_sapp.py
import unittest
import gzip
import json
import sys

import sapp

from types import SimpleNamespace

class TestSearchMethod(unittest.TestCase):
    def test_cities(self):
        print ("running test_cities...")
        with gzip.open('data.json.gz', 'rb') as db:
            cities = json.load(db)
        
        states = set()
        errors, count = 0, 0
        for city in cities:
            c = SimpleNamespace(**city)
            states.add(c.uf)
            r = sapp._search(c.latitude, c.longitude)
            if r is None:
                errors += 1
            else:
                if r["found"]:
                    if c.nome_municipio == r["city"] and c.uf == r["state"]:
                        count += 1
                    else:
                        errors += 1
                else:
                    errors += 1
        
        self.assertTrue(errors==0)
        self.assertTrue(count==5540)
        self.assertTrue(len(states)==27)

    def test_states(self):
        print ("running test_states...")
        with gzip.open('capitais.json.gz', 'rb') as db:
            all_br_states = json.load(db)

        with gzip.open('data.json.gz', 'rb') as db:
            cities = json.load(db)

        all_br_states = [x['sigla'] for x in all_br_states]
        all_states = [x['uf'] for x in cities if x['capital']]
        
        self.assertTrue(len(all_states)==27)
        self.assertTrue(len(all_br_states)==27)
        
        results = all(elm in all_states for elm in all_br_states)
        self.assertTrue(results)

    def test_capitals(self):
        print ("running test_capitals...")
        with gzip.open('capitais.json.gz', 'rb') as db:
            all_capitals = json.load(db)

        with gzip.open('data.json.gz', 'rb') as db:
            cities = json.load(db)
        
        cities = [x for x in cities if x['capital']]

        found = []
        for capital in all_capitals:
            c = SimpleNamespace(**capital)
            for city in cities:
                d = SimpleNamespace(**city)
                if({d.nome_municipio, d.uf, d.estado} & {c.capital, c.sigla, c.estado}):
                    found.append(capital)
                    break

        self.assertTrue(len(found)==27)


if __name__ == '__main__':
    unittest.main()

convert.py
"""
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; version 2 of the License.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License 
along with this program. If not, see <http://www.gnu.org/licenses/>.

Created on 2019-05-11
By P&D Aplicativos
"""

import gzip
import pickle
import sys

from fastkml import kml

E = {
    "SP": 0,
    "MG": 1,
    "RJ": 2,
    "DF": 3,
    "PR": 4,
    "BA": 5,
    "CE": 6,
    "SC": 7,
    "RS": 8,
    "ES": 9,
    "PA": 10,
    "PE": 11,
    "TO": 12,
    "AM": 13,
    "PB": 14,
    "MS": 15,
    "GO": 16,
    "AL": 17,
    "MA": 18,
    "PI": 19,
    "RN": 20,
    "SE": 21,
    "MT": 22,
    "RR": 23,
    "AP": 24,
    "RO": 25,
    "AC": 26,
}

C = {
    "São Paulo": (0, "SP"),
    "Belo Horizonte": (1, "MG"),
    "Rio de Janeiro": (2, "RJ"),
    "Brasília": (3, "DF"),
    "Curitiba": (4, "PR"),
    "Salvador": (5, "BA"),
    "Fortaleza": (6, "CE"),
    "Florianópolis": (7, "SC"),
    "Porto Alegre": (8, "RS"),
    "Vitória": (9, "ES"),
    "Belém": (10, "PA"),
    "Recife": (11, "PE"),
    "Palmas": (12, "TO"),
    "Manaus": (13, "AM"),
    "João Pessoa": (14, "PB"),
    "Campo Grande": (15, "MS"),
    "Goiânia": (16, "GO"),
    "Alagoas": (17, "AL"),
    "São Luis": (18, "MA"),
    "Teresina": (19, "PI"),
    "Natal": (20, "RN"),
    "Aracaju": (21, "SE"),
    "Cuiabá": (22, "MT"),
    "Boa Vista": (23, "RR"),
    "Macapá": (24, "AP"),
    "Porto Velho": (25, "RO"),
    "Rio Branco": (26, "AC"),
}

with gzip.open('data.kml.gz', 'rb') as k:
    doc = k.read()

k = kml.KML()
k.from_string(doc)

f = list(k.features())
d = list(f[0].features())

def helper(v):
    t = v.extended_data.elements[0].data
    h = {}
    for e in list(t):
        h[e["name"]] = e["value"]
    if "NOME_MUNIC" not in h:
        return (999, "")
    if h["NOME_MUNIC"] in C:
        s = C[h["NOME_MUNIC"]]
        if h["SIGLA"] == s[1]:
            return (s[0], h["NOME_MUNIC"])
    elif h["SIGLA"] in E:
        s = E[h["SIGLA"]] + len(C) + 1
        return (s, h["SIGLA"])
    return (99, h["NOME_MUNIC"])

d.sort(key=helper)

with gzip.open("data.dat.gz", "wb") as f:
    pickle.dump(d, f)



sapp.py
"""
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; version 2 of the License.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License 
along with this program. If not, see <http://www.gnu.org/licenses/>.

Created on 2019-05-11
By P&D Aplicativos
"""

import copy
import functools
import json
import geohash
import gzip
import multiprocessing
import pickle
import socket
import time
import uuid
import sys

from collections import namedtuple, Counter
from datetime import datetime

from sanic import Sanic
from sanic.response import json as jsonify
from sanic.exceptions import SanicException
from sanic_cors import CORS, cross_origin

from fastkml import kml
from shapely.geometry import Point

from types import SimpleNamespace


with gzip.open('data.dat.gz', 'rb') as k:
    d = pickle.load(k)

print(sys.version_info)
C = { "global": Counter(), "stats": Counter() }
G = SimpleNamespace(**{"config": None})

with open('sapp.json', 'rt', encoding='utf-8') as j:
    G.config = json.load(j)
    G.config = SimpleNamespace(**G.config)
    G.config.tokens = set(G.config.tokens)
    for token in G.config.tokens:
        if token not in C:
            C["stats"][token[:6]] = Counter()
    DG = copy.deepcopy(G)
    DG.config.tokens = [t[:6] for t in G.config.tokens]
    print(DG)


@functools.lru_cache(maxsize=G.config.maxsize)
def search(ghash):
    latitude, longitude = geohash.decode(ghash)
    return _search(latitude, longitude)


def _search(latitude, longitude):
    p = Point(longitude, latitude)
    for i, c in enumerate(d):
        g = c.geometry
        if p.within(g):
            t = list(c.extended_data.elements[0].data)
            h = {e["name"]:e["value"] for e in t if e["name"] in ["SIGLA", "NOME_MUNIC"]}
            return {"city": h["NOME_MUNIC"], "state": h["SIGLA"], "found": True}
    return None


app = Sanic()
app.config.KEEP_ALIVE = False


@app.middleware('request')
async def before_request(request):
    g = namedtuple("g", ["start_time", "request_id"])
    request["g"] = g(datetime.now(), str(uuid.uuid4()))


@app.middleware('response')
async def after_request(request, response):
    g = request["g"]
    dt = datetime.now() - g.start_time
    ms = str(round((dt.days * 24 * 60 * 60 + dt.seconds) * 1000 + dt.microseconds / 1000.0))
    response.headers['X-Profile'] = ms
    response.headers['X-Req'] = g.request_id
    response.headers['X-Version'] = G.config.version
    response.headers['X-Id'] = f"{request.ip}:{request.port}"


@app.route('/check')
async def check(request):
    return jsonify(C)


@app.route('/', methods=['GET'])
@cross_origin(app, origins=G.config.origins, automatic_options=True)
async def index(request):
    token = request.headers.get("X-Token")
    if token not in G.config.tokens:
        C["global"].update({"401": 1})
        raise SanicException("Unauthorized", status_code=401)
    try:
        latitude = float(request.args.get("lat", default=None))
        longitude = float(request.args.get("lon", default=None))
        ghash = geohash.encode(latitude, longitude, G.config.precision)
    except:
        C["stats"][token[:6]].update({"400": 1})
        raise SanicException("Bad Request", status_code=400)
    try:
        data = search(ghash)
        if data is None:
            C["stats"][token[:6]].update({"404": 1})
            return jsonify(G.config.fallback)
        else:
            C["stats"][token[:6]].update({"200": 1})
            return jsonify(data)
    except:
        C["stats"][token[:6]].update({"500": 1})
        return jsonify(G.config.fallback)


if __name__ == '__main__':
    if len(sys.argv) > 1 and sys.argv[1] == "--console":
        pass
    else:
        app.run(host='0.0.0.0', port=8080, workers=multiprocessing.cpu_count())